packages/dataviews/src/components/dataviews/index.tsx

```diff
@@ -177,11 +181,19 @@ function DataViews< Item >( {
 		}
 	}
 	const _fields = useMemo( () => normalizeFields( fields ), [ fields ] );
+	const isPicker = Boolean( picking );
 	const _selection = useMemo( () => {
-		return selection.filter( ( id ) =>
-			data.some( ( item ) => getItemId( item ) === id )
-		);
-	}, [ selection, data, getItemId ] );
+		// The selection is filtered here to known `data` to ensure items that
+		// may have been deleted are not present in the selection.
+		// For picker mode, this prevents selecting items that might be on
+		// different pages, since `data` only contains the current paginated view.
+		// Hence the filtering is only done when `isPicker` is `false`.
```

Comment:
Tiny nit: I found the `For picker mode...` sentence a bit ambiguous on a first read, so wondered if we reorder the statements, if it might help clarity:

```suggestion
		// The filtering is only done when `isPicker` is `false`, because
		// in picker mode, this prevents selecting items that might be on
		// different pages, since `data` only contains the current paginated view.
```